### ISSUE TITLE
v3.0.0 `await res.json()` is `unknown`

### ISSUE THREAD

**Issue by cobaltt7** [node-fetch/node-fetch#1262]
this is unusable in typescript. v3.0.0 types the return value of `await res.json()` as `unknown` which means "this is dangerous info, you have no idea what it is. don't touch it to avoid runtime errors." but the point of a `fetch` request is to get the information back. you _have_ to touch it. the type should stay `any`.

**Reproduction**

1. use node-fetch in a typescript project
2. try to build
3. see error

**Expected behavior**

for the build to succeed

**Comment by RedGuy12** [comment]
this can be fixed by changing this line
https://github.com/node-fetch/node-fetch/blob/2603c67f1840fb4a43e0837099a3b2725e8f615e/%40types/index.d.ts#L110

**Comment by ImRodry** [comment]
YES! The package from DefinitelyTyped had this as any but when 3.0 made the change to add type declarations they set it to unknown for reasons that were explained in that PR (will try to find it and edit this if I do). This makes our lives way harder for no reason at all since unknown types, like you mentioned, are for things you shouldn't be touching, which is definitely not the case here. I'd say you should open a PR tbh

**Comment by efstathiosntonas** [comment]
on fast workaround would be (if this is never get's reversed to `any`)

```
const res = (await response.json()) as any;

```

**Comment by ImRodry** [comment]
TS types the result of .json() as Promise<any> for browsers so if anything it should follow the types made by TS itself, and not make up stuff because you think it makes sense but you actually just misread documentation

**Comment by LinusU** [comment]
Unless TypeScript provides a native json type (https://github.com/microsoft/TypeScript/issues/1897), I think that `unknown` is absolutely preferred over `any`.

If you have it return `any`, you don't get any type safety at all and the benefits of using TypeScript is being negated.

It's very easy to cast the value to whatever type you want, and then it's very clear in the code that you are saying to the compiler "trust me, this data will always have the type I'm saying":

```ts
const data = await res.json() as any as Foobar
```

**Comment by Richienb** [comment]
We can probably just type out a normal JSON object (see https://github.com/sindresorhus/type-fest/blob/2783a0890925bc180a406d0e34290578812b4864/source/basic.d.ts#L15-L45)

**Comment by LinusU** [comment]
@Richienb hmm, the documentation for the type you linked says this though 🤔 

> Don't use this as a direct return type as the user would have to double-cast it: `jsonObject as unknown as CustomResponse`

**Comment by Richienb** [comment]
I see. In which case, we should just be following whatever is set in the official window.fetch typings (if they exist).

**Comment by LinusU** [comment]
Some relevant discussions:

- https://github.com/nodejs/undici/pull/911/files#r680432742
- https://github.com/microsoft/TypeScript/issues/33037

Personally, I feel that `unknown` provides the least amount of danger of shooting oneself in the foot, which I think it most important. But I can see us going with `any` to match `dom.d.ts`. Do note that these types was added to `dom.d.ts` before the `unknown` type was added to TypeScript, so at the time they didn't had a choice...

**Comment by ImRodry** [comment]
> Some relevant discussions:
> 
> * https://github.com/nodejs/undici/pull/911/files#r680432742
> * [Improved type definitions for JSON api responses microsoft/TypeScript#33037](https://github.com/microsoft/TypeScript/issues/33037)
> 
> Personally, I feel that `unknown` provides the least amount of danger of shooting oneself in the foot, which I think it most important. But I can see us going with `any` to match `dom.d.ts`. Do note that these types was added to `dom.d.ts` before the `unknown` type was added to TypeScript, so at the time they didn't had a choice...

I have multiple instances in my code where I don’t need to make an interface for the JSON I get from the fetch call and only do so later on in a property of that JSON. For example, why would I need to make an interface for an object that looks like this
```json
{
    “data”: {
        …
    }
}
```
When only the type inside `data` matters to me? I even tried doing this but I just ended up with tons of useless interfaces that mean absolutely nothing and present no advantage at all. By giving us the any type you’re trusting that we know what we’re doing, instead of saying **you have to type this, otherwise we won’t let you go any further**

You also mentioned TS didn’t have a choice when the window.fetch types were added but the unknown type has been a thing for quite a while and they’ve had plenty of time to change it. If they didn’t then it’s because they didn’t see a reason to do that, and I don’t think you should have any saying on that when your package just ports something from the browser to node. This is not node-fetch anymore, this is something you made up yourselves.

**Comment by RedGuy12** [comment]
Here's a type that allows a bit more flexibility:

```ts
interface json {
[x:string]:unknown|json
}
```


________________________________

**Comment by ChromeQ** [comment]
I was already casting my response like so: 
```ts
const response = await fetch(url, { headers });
const json: MySpecialResponseType = await response.json();
```
Since any was pretty much useless for type safety so I think switching to `unknown` is the right one. The simple solution is to simply cast to your wanted type like this:
```ts
const response = await fetch(url, { headers });
const json = (await response.json()) as MySpecialResponseType;
```
The more "robust" and "correct" solution would be to use a type guard check to 100% make sure your response actually is the type you expect, this is invaluable when interacting with 3rd party endpoints rather than your own I guess.

**Comment by ImRodry** [comment]
@ChromeQ that is all fine when the type that you get from the fetch method actually matters to you. Please refer to my message above, we shouldn’t be forced to type cast something we don’t need to use

**Comment by ChromeQ** [comment]
If the type doesn't matter to you then let it stay as unknown. But the fact remains, there is no way to be absolutely sure what the server returned at compile time, that is why there are only 2 ways out: cast to tell TS you know better and you know you trust the response (which you had to do anyway with any type) or, you can have runtime checks with type guards which TS can understand and help you out at compile time.
Unknown type is a better choice for the response than any. I'm just offering people a quick solution to the error

**Comment by ChromeQ** [comment]
> I even tried doing this but I just ended up with tons of useless interfaces that mean absolutely nothing and present no advantage at all

I see your issue as I hate useless interfaces and repetition, but I think you can use a generic type quite efficiently to get around writing "tons of useless interfaces" where you cast your json to this new generic type with the interface you care about as the generic which is the interface without the `"data"` wrapper.

Happy to provide an example if you share a sample of your code :)

**Comment by ImRodry** [comment]
> If the type doesn't matter to you then let it stay as unknown. But the fact remains, there is no way to be absolutely sure what the server returned at compile time, that is why there are only 2 ways out: cast to tell TS you know better and you know you trust the response (which you had to do anyway with any type) or, you can have runtime checks with type guards which TS can understand and help you out at compile time.
> Unknown type is a better choice for the response than any. I'm just offering people a quick solution to the error

You can't simply "let it stay as unknown" because that would mean you cannot do anything with the JSON, nor access any of its properties. In the case I described above I only need to access 1 property of the JSON which is then properly typed, but not the JSON itself. I only used node-fetch to send requests to public APIs in my code so yes I can be absolutely sure of what the API is going to return if it's all properly documented. Either way, setting the type to unknown is just taking away the flexibility we had before and forcing us to do something that we may not want, and even TS itself does not want.

> > I even tried doing this but I just ended up with tons of useless interfaces that mean absolutely nothing and present no advantage at all
> 
> I see your issue as I hate useless interfaces and repetition, but I think you can use a generic type quite efficiently to get around writing "tons of useless interfaces" where you cast your json to this new generic type with the interface you care about as the generic which is the interface without the `"data"` wrapper.
> 
> Happy to provide an example if you share a sample of your code 🙂

I believe someone already showed an example of this generic interface above and I also showed an example of a JSON object I get in a specific part of my code so that's fine, but wouldn't this generic type put us back with the same "problem" you described with the any type? I mean if you use this you're essentially saying any property on your JSON can be anything, which isn't type safe and is pretty much useless. Any workaround to this change is essentially a waste of time and code, even if it doesn't make it into the compiled code, it's still making our lives much much harder for no reason at all

**Comment by LinusU** [comment]
> You can't simply "let it stay as unknown" because that would mean you cannot do anything with the JSON, nor access any of its properties. In the case I described above I only need to access 1 property of the JSON which is then properly typed, but not the JSON itself. 

In that case, can you do something like this?

```ts
const data = await res.json() as { data: Foobar }
```

That way you don't have to create an additional type...

> Either way, setting the type to unknown is just taking away the flexibility we had before and forcing us to do something that we may not want, and even TS itself does not want.

You can always do `as any` to tell TS that it's "anything". I don't really understand what you mean with "something that we may not want" 🤔

**Comment by ChromeQ** [comment]
When you called .json you got a type of `any`, at some point you must have type narrowed or casted it to something useful you want to use.
```ts
const value: any = "I'm a string";
value.call.a.function.that.does.not.exist(); // this is ok when it is `any`
```
The reason your code worked before is either because you relied on the TS escape-hatch or luck, or because you "send requests to public APIs in my code so yes I can be absolutely sure of what the API is going to return" which I would argue is lucky since just because an API is public does not mean it _certainly_ will return what it says it will, even "if it's all properly documented"
The point is, the TS compiler **cannot** know for sure what the response from the server will be. Which is why you have to overrule TS by casting, or you add some runtime checks in the form of typeguards which TS will understand and help you at compile/code writing stage

**Comment by martijndeh** [comment]
I don't mind defaulting to `unknown` instead of `any`, but can we have something like this `json<T = unknown>(): Promise<T>;` instead?

**Comment by ChromeQ** [comment]
Absolutely, I was trying to use it like this and was surprised it wasn't possible.

**Comment by LinusU** [comment]
> [...] but can we have something like this `json<T = unknown>(): Promise<T>;` instead?

I don't think that we should do this. It was discussed in both of the links that I posted above, but the argument basically boils down to that it's a hidden forced typecast. I think that it's easy to accidentally miss that you are unsafely changing the type of the data, when no compile time or runtime check is actually being done on it.

**Comment by martijndeh** [comment]
Aha, I see. Up to you of course. In my opinion it's a bit more ergonomic to provide a type argument in the cases where you opt out of any runtime checks anyway — even if DefinitelyTyped thinks this is a common mistake. ;-)

```ts
// Explicit type cast
const json1 = (await response.json()) as MyJson;
// Alternative type cast, not much better
const json2 = await (response.json() as Promise<MyJson>);
// Doesn't work anymore, 'unknown' is not assignable to type 'MyJson'
const json3: MyJson = await response.json();
// So pretty ;-)
const json4 = await response.json<MyJson>();
```

**Comment by tatemz** [comment]
I prefer to be as strict as possible and avoid using `any`. Instead I use `unknown` or `never`. [User defined type guards](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) are how I: 

1. Validate the response JSON is indeed what my business logic needs (throwing an exception if it isn't)
2. Narrowing the type from `response.json()`

As @martijndeh has indicated, if you don't want a runtime check, the `response.json<MyType>` is the way to go.

**Comment by OmgImAlexis** [comment]
> > [...] but can we have something like this `json<T = unknown>(): Promise<T>;` instead?
> 
> I don't think that we should do this. It was discussed in both of the links that I posted above, but the argument basically boils down to that it's a hidden forced typecast. I think that it's easy to accidentally miss that you are unsafely changing the type of the data, when no compile time or runtime check is actually being done on it.
> 
> edit to add:
> 
> DefinitelyTyped also lists this as something you shouldn't do under their [common misstakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes)

There's little to no difference between casting from unknown to a type vs using a generic with unknown as the default.

**Comment by Bessonov** [comment]
From my point of view, it's wrong and you make it more complicated than it should be. But let's me explain this.

First of all, I'm type-safety advocate. But fetch is very special, because it doesn't have any type-safety mechanics. You just get an _arbitrary_ blob of _something_. You can't be even sure that you can call `.json()` at all. In this situation you **can't** type the body. There is no way. Using `unknown` instead of `any` do only one thing: it forces to give some _imaginary_ type. All `json() as MyData` or `json<MyData>()` has nothing to do with reality.

To make the body type safe (and after checking that it is truly json body) you _must_ check it at runtime. For example with using:
- [OpenAPI](https://www.openapis.org/) or [JSON Schema](https://json-schema.org/)
- [ajv](https://github.com/ajv-validator/ajv)
- [assertion functions](https://2ality.com/2020/06/type-guards-assertion-functions-typescript.html#asserting-a-boolean-argument%3A-asserts-%C2%ABcond%C2%BB)

And because of that there is no benefit trying make `json()` type safe with typescript only. Therefore, I vote for `json<T = unknown>(): Promise<T>` just because it is more _ergonomic_ and more _expected_. But I was even fine with `any` for this special case - there was no surprises with fetch.

**Comment by tatemz** [comment]
Got notified of @Bessonov's comment today and came back to this issue in 2022 to say that I learned about assertion functions ([here's another link about it](https://dev.to/smeijer/typescript-type-assertions-4klf)) 2 months ago and it's useful if you don't plan on using runtime validation. I have never seen official TS documentation on the `asserts` keyword 🤷 but it's nifty and useful when you trust the shape of a thing.

When I don't trust the shape of a thing, I still use runtime validation. When I do, I have been using [`runtypes`](https://www.npmjs.com/package/runtypes) to type and validate an API response object more conveniently! (ty @guyellis for showing me that package 😄)

I'd also vote for `json<T = unknown>(): Promise<T>` so that I'm forced to deal with the fallout of my own type-laziness 😅. 

Instead of `unknown`, I'd even be cool with `Array<unknown> | Record<string | number, unknown>` as that is closer to what JSON is allowed to be. Ultimately, this comes down to the `JSON.parse()` method which is the source of my temptation to be type-lazy 😈

**Comment by tatemz** [comment]
> Do you think that typing `await response.json<Foo>()` is so much more convenient than typing `await response.json() as Foo`?

@LinusU I don't have a preference, but that is not the crux of the matter for me. I would prefer `await response.json()` to have the return type of `unknown` instead of `any`. At the end of the day, all this is moot IMO because I've learned to be more and more type-safe as I grow as a developer. If anything, this issue and thread will continue to educate onlooking devs that are recently making the switch out of JS or making the switch to stronger `tsconfig.json` configurations.

`node-fetch` has an opportunity to build a more type-safe library and naturally educate devs (like myself) through its type system. In the meantime, we can add functions like this to build stronger type systems in our applications. 

```ts
const getJSONFromResponse = <T = unknown>(response: Response): Promise<T> => response.json();
```

**Comment by Bessonov** [comment]
@tatemz 

> and it's useful if you don't plan on using runtime validation

I've corrected my comment. I meant the combination of OpenAPI definition, ajv and assertion function. I use this combination in all places:
- validation of http-requests bodies coming outside aka inside request handler
- validation of events coming from kafka
- validation of fetch responses.

For example:
```typescript
const payload = await parseJSON(ctx, req) // <-- get json from a request, but works for fetch responses in the same way
ensureUserProfileSearchRoot(ctx, payload) // <-- assertion function with runtime validation against OpenAPI definition - created any way
const userProfileSearchRoot = UserProfileSearchRootFromJSON(payload) // <-- generated by [OpenAPI generator](https://github.com/OpenAPITools/openapi-generator) from the same OpenAPI definition
return searchUserProfile(ctx, userProfileSearchRoot) // <-- type safe usage
```

@LinusU 

> Do you think that typing await response.json<Foo>() is so much more convenient than typing await response.json() as Foo?

I see two points:
1. For me it's more convenient. But, well, I wouldn't die by typing ` as ` (4 chars) instead of `<>` (2 chars) :)
2. It's more expected. Like some others in this topic, I tried to type `json<MyData>()` before I was surprised by the definition.

> The benefit of having the latter is that organisations that really care about type safety can use lints to disallow unsafe type casting, whereas if we allow the former there is no way to statically infer that unsafe type casting is happening in the code.

Damn, I agree with it. It's a very strong argument for me, even I'm not convinced that it is the majority of ts users.

**Comment by casey6** [comment]
Like others I came here because I was very surprised at the current behavior.

I don't think it's fair to categorize the desire for a generic in this situation as not really caring about type safety.

By its own nature `fetch` is designed for consuming other APIs. If an API says they will return some JSON in the form of a `Foo` when you make a `GET /foo` request, then it's safe to expect `response.json()` to return a `Promise<Foo>`.

**Comment by LinusU** [comment]
@casey6 is there any downside for you to use `as Foo` instead of `<Foo>`?

**Comment by LinusU** [comment]
I feel like Axios and Got are high level client libraries which might explain why they chose that route. The goal with node-fetch is to follow the browser fetch as closely as possible, and in the bundled TypeScript bindings `json()` is typed as returning `any`:

https://github.com/microsoft/TypeScript/blob/59238dbf90f7fb002357851940ced2ae42800e69/lib/lib.dom.d.ts#L2448

The reason we are going with `unknown` instead of `any` is discussed above. But if we are going to switch to allow `<Foo>` I think that this needs to first land upstream in TypeScript. For the reasons I have laid out above however, I don't think that TypeScript will accept such a change.

If we introduced a generic type-parameter here we would conflict with the `dom` typings. `unknown` doesn't to this as it's assignable to `any`. This means that projects targeting both the browser and node could have problems...

**Comment by casey6** [comment]
I was going to comment on that earlier but didn't want to get too long-winded. The spirit of that suggestion is that the inputs of your function should affect the output. The proposed implementation here might not enforce that, but I do think that's the intent with normal usage. `fetch('/foos')` returns `Array<Foo>`, `fetch('/foos/{id}')` returns `Foo`, etc. 

I could imagine an implementation that goes a bit further where the generic is added to `fetch` itself, and includes both the url and json return type. A simplistic version:

```ts
type FetchRequest = {
    url: string,
    json: unknown
}

declare function fetch<F extends FetchRequest = FetchRequest>(url: F['url']): Promise<{ json: () => Promise<F['json']> }>

type FooRequest = {
    url: '/foo'
    json: Foo
}

const response = await fetch('/foo')
const data = await response.json() // data is unknown

const response = await fetch<FooRequest>('/foo')
const data = await response.json() // data is a Foo

const response = await fetch<FooRequest>('/bar') // error
```

That would satisfy the DefinitelyTyped advice. It's impractical for a number of reasons, so I'm not actually suggesting it. But I mention it to show the _intent_ of a developer using a generic in this situation. It's not about forcing types, it's about modeling responses.

I think it's important to have perspective too. Using `any` is a far more egregious "mistake". Yet the official typescript definitions use that.

**Comment by denexapp** [comment]
I think it's pretty safe to use `object` instead of `unknown` by default.

**Comment by jimmywarting** [comment]
> use object

JSON can also return only strings, booleans, numbers also. JSON isn't always a object

**Comment by denexapp** [comment]
Good point. How about `object | string | boolean | number` then? While its not as narrow as it can possibly be, it's more specific than `unknown`.

**Comment by adrian-gierakowski** [comment]
Any input you get from outside world is unknown and unsafe to use without validation, therefore `unknown` is the correct type here (you could use Json as proposed above, but there is not much value in it). If you want to do anything with the value, use [type narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html) or schema validation libraries like [zod](https://github.com/colinhacks/zod) or [@effect/schema](https://github.com/effect-ts/effect/tree/main/packages/schema).

You are free to cast the value to whatever you want but you are doing it at your own risk.
